### PR TITLE
[SMT backend] Improve Overflow Handling in ESBMC

### LINKED ISSUE #2337: ESBMC boolector crash
For the following code ESBMC 7.8.1 is crashing with

```
ERROR: Boolector error encountered
[boolector] boolector_eq: nodes must have equal sorts

Abort trap: 6
```

ESBMC command:  `esbmc esbmc_crash.c --overflow`

CODE:
```c
#include <stdio.h>
#include <stdlib.h>

int maximum(int x, int y) {
    return x ^ ((x ^ y) & -(x < y));
}
int main() {
   
    unsigned int x = 15, y = 20;
    printf("    [*] The maximum of %d and %d is %d\n", 32, 56, maximum(32, 56));
    return 0;
}
```

### PR DESCRIPTION
**Pull request by lucasccordeiro**
Fixes https://github.com/esbmc/esbmc/issues/2337.

This PR improves overflow detection in ESBMC by refining the implementation of `overflow_neg` in `smt_overflow.cpp`. Additionally, it adds regression tests to ensure correct behavior when checking for arithmetic overflow, specifically negation-related overflows.

### CODE REVIEW

**Comment by intrigus-lgtm** [comment]
Does this also fix https://github.com/esbmc/esbmc/issues/1377 ?

File: src/solvers/smt/smt_overflow.cpp [hunk 1]
```diff
@@ -204,12 +204,20 @@ smt_astt smt_convt::overflow_neg(const expr2tc &expr)
...
 
-  // Single failure mode: MIN_INT can't be neg'd
+  // Extract operand
   const overflow_neg2t &neg = to_overflow_neg2t(expr);
   unsigned int width = neg.operand->type->get_width();
 
-  expr2tc min_int =
-    constant_int2tc(neg.operand->type, -BigInt::power2(width - 1));
-  expr2tc val = equality2tc(neg.operand, min_int);
+  // Ensure operand is an integer (convert boolean to signed integer if needed)
```
Comment (intrigus-lgtm) [inline src/solvers/smt/smt_overflow.cpp]: Why convert to a signed integer and not an unsigned integer instead?

**Comment by lucasccordeiro** [comment]
> Does this also fix https://github.com/esbmc/esbmc/issues/1377 ?

@intrigus-lgtm: This PR does not fix this issue. Let me reduce the program and fix this issue in another PR.

**Review by intrigus-lgtm — COMMENTED** [review]
This still feels a bit like a hack.
E.g. see https://github.com/esbmc/esbmc/issues/1379#issuecomment-1739036916

**Comment by XLiZHI** [comment]
I think `github_2337_2`, `github_2337_3`, `github_2337_4`, and `github_2337_5` are not related to this PR's change. The cause of the error is mainly the bool operation ( a > b ).

File: src/solvers/smt/smt_overflow.cpp (hunk 1 above)
Comment (XLiZHI) [reply src/solvers/smt/smt_overflow.cpp:211]: I guess you mean the option?
`{"unsigned-overflow-check", NULL, "enable arithmetic over- and underflow check for unsigned integers"}`

Comment (lucasccordeiro) [reply src/solvers/smt/smt_overflow.cpp]: The conversion to a signed integer is done to handle the overflow caused by negating the smallest signed integer (INT_MIN). If we use an unsigned integer, there would be no overflow for the negation because of the wrapping behavior.

**Comment by lucasccordeiro** [comment]
> This still feels a bit like a hack.

@intrigus-lgtm: Do you think you could submit an alternative PR? :-)

As @mikhailramalho pointed out:

"I agree that this is unsatisfying, we should do everything following the
C/C++ standard up to before converting to SMT and then convert the data to
the format that SMT expects, but I think this will be a huge project."

**Comment by lucasccordeiro** [comment]
> I think github_2337_2, github_2337_3, github_2337_4, and github_2337_5 are not related to this PR's change. The cause of the error is mainly the bool operation ( a > b ).

@XLiZHI: I agree with you, but I don't think it harms to add these test cases to check the other cases in the overflow method ;-)

File: src/solvers/smt/smt_overflow.cpp [hunk 2]
```diff
@@ -204,12 +204,20 @@ smt_astt smt_convt::overflow_neg(const expr2tc &expr)
...
+  expr2tc operand = is_bool_type(neg.operand->type)
```
Comment (rafaelsamenezes) [inline src/solvers/smt/smt_overflow.cpp]: Can't you always do the typecast? Our simplifier should be smart enough to handle double typecast.

**Review by rafaelsamenezes — APPROVED** [review]
I would also suggest creating an umbrella issue so we do not lose track of these typecast hacks.

File: src/solvers/smt/smt_overflow.cpp (hunk 2 above)
Comment (lucasccordeiro) [reply src/solvers/smt/smt_overflow.cpp]: Implemented as suggested.

**Review by XLiZHI — APPROVED** [review]

**Comment by lucasccordeiro** [comment]
I have just created this umbrella issue: https://github.com/esbmc/esbmc/issues/2339

**Comment by intrigus-lgtm** [comment]
Unrelated to this PR, but I'd say the code should simply ignore unsignedbvs:

`esbmc --overflow-check --unsigned-overflow-check`
```c
int main() {
unsigned int a = 0x80000000;
return -a;
}
```

```
State 3 file foo.c line 3 column 1 function main thread 0
----------------------------------------------------
Violated property:
  file foo.c line 3 column 1 function main
  arithmetic overflow on neg
  !overflow("unary-", a)


VERIFICATION FAILED
```

**Comment by lucasccordeiro** [comment]
Thanks, @intrigus-lgtm. 

I have just created this PR to fix this issue: https://github.com/esbmc/esbmc/pull/2340.
